### PR TITLE
ray dashboard command for clusters

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This introduces a `ray dashboard` command, which sets up port forwarding so that a remote cluster's dashboard an be viewed locally. In particular, the dashboard should be accessible at localhost:8265, or the next available port as long as the command is running. 

## Related issue number

N/A
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/autoscaler/updater.py [hunk 1]
```diff
@@ -58,13 +58,26 @@ def run(self,
...
-            # running the actual command. This is a little ugly, but it should
-            # work in most scenarios and nothing should go very wrong if the
-            # command starts running before the port forward starts.
-            time.sleep(1)
+            try:
+                # Give port-forward a chance to connect before running the
+                # actual command. If port forward fails (and returns
+                # immediately) we failed to port forward and should abort the
+                # command
+                port_forward_process.wait(5)
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]: This seems like a very long time to wait before launching the other process. This should fail relatively fast, no?

File: python/ray/autoscaler/updater.py [hunk 2]
```diff
@@ -58,13 +58,26 @@ def run(self,
...
+            except subprocess.TimeoutExpired:
+                # Lasting for more than a second is a good thing
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]: comment doesn't match the actual timeout

File: python/ray/autoscaler/updater.py [hunk 3]
```diff
@@ -173,6 +186,9 @@ def get_default_ssh_options(self, connect_timeout):
             ("ControlPath", "{}/%C".format(self.ssh_control_path)),
             ("ControlPersist", "10s"),
             ("IdentitiesOnly", "yes"),
+            ("ExitOnForwardFailure", "yes"),
+            ("ServerAliveInterval", 5),
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]: It would be nice to drop a comment on these for why we need them

File: python/ray/scripts/scripts.py [hunk 4]
```diff
@@ -60,6 +60,80 @@ def cli(logging_level, logging_format):
...
+@click.option(
+    "--port",
+    "-p",
+    required=False,
+    type=int,
+    default=8265,
+    help="The local port to forward to the dashboard")
+def dashboard(cluster_config_file, cluster_name, port):
+    # Sleeping in a loop is preferable to `sleep infinity` because the latter
+    # only works on linux
```
Comment (edoakes) [inline python/ray/scripts/scripts.py]:
```suggestion
    # only works on linux.
```

File: python/ray/scripts/scripts.py [hunk 5]
```diff
@@ -60,6 +60,80 @@ def cli(logging_level, logging_format):
...
+    cmd = "while true; do sleep 86400; done"
+    remote_port = 8265
+    if port:
+        dashboard_port = port
+    else:
+        dashboard_port = remote_port
+
+    port_taken = True
+
+    # Figure out what's the first port we can use
```
Comment (edoakes) [inline python/ray/scripts/scripts.py]:
```suggestion
    # Find the first open port sequentially from `remote_port`.
```

File: python/ray/scripts/scripts.py [hunk 6]
```diff
@@ -60,6 +60,80 @@ def cli(logging_level, logging_format):
...
+    while port_taken:
+        try:
+            port_forward = [
+                (dashboard_port, remote_port),
+            ]
+            click.echo(
+                "Attempting to establish dashboard at localhost:{}".format(
+                    port_forward[0][0]))
+            # We want to probe with a no-op that returns quickly to avoid
+            # exceptions caused by network errors
```
Comment (edoakes) [inline python/ray/scripts/scripts.py]:
```suggestion
            # exceptions caused by network errors.
```

File: python/ray/scripts/scripts.py [hunk 7]
```diff
@@ -60,6 +60,80 @@ def cli(logging_level, logging_format):
...
+            exec_cluster(
+                cluster_config_file,
+                ":",
+                docker=False,
```
Comment (edoakes) [inline python/ray/scripts/scripts.py]: No need to explicitly set default options here

**Review by edoakes — COMMENTED** [review]
Would also be good to add this somewhere in the documentation (though I'm not sure if we have a dashboard page ATM)

File: python/ray/autoscaler/commands.py [hunk 8]
```diff
@@ -415,8 +415,11 @@ def wrap_docker(command):
             attach_info = "Use `{}` to check on command status.".format(
                 attach_command)
             logger.info(attach_info)
-    finally:
+    except Exception as e:
         provider.cleanup()
+        raise e
+
+    provider.cleanup()
```
Comment (robertnishihara) [inline python/ray/autoscaler/commands.py]: Out of curiosity, what difference does this change make? It looks equivalent to me.

File: python/ray/scripts/scripts.py (hunk 7 above)
Comment (wuisawesome) [reply python/ray/scripts/scripts.py]: `exec_cluster` currently doesn't have default arguments, are you suggesting that I give it default arguments?

File: python/ray/autoscaler/commands.py (hunk 8 above)
Comment (wuisawesome) [reply python/ray/autoscaler/commands.py]: Good catch, I forgot to revert this after debugging.

File: python/ray/scripts/scripts.py (hunk 7 above)
Comment (edoakes) [reply python/ray/scripts/scripts.py]: Oh I see. For clarity we should avoid using `kwarg` syntax when there aren't defaults defined. Defining default arguments probably makes sense, the current code that calls this is quite ugly.

**Comment by wuisawesome** [comment]
Documentation: Screen Shot 2020-02-04 at 3 02 56 PM

File: doc/source/autoscaling.rst [hunk 9]
```diff
@@ -211,6 +211,8 @@ Port-forwarding applications
 
 If you want to run applications on the cluster that are accessible from a web browser (e.g., Jupyter notebook), you can use the ``--port-forward`` option for ``ray exec``. The local port opened is the same as the remote port.
 
+Note: For Kubernetes clusters, the ``port-forward`` option cannot be used while executing a command. To port forward and run a command you need to call ``ray exec`` twice.
```
Comment (edoakes) [inline doc/source/autoscaling.rst]:
```suggestion
Note: For Kubernetes clusters, the ``port-forward`` option cannot be used while executing a command. To port forward and run a command you need to call ``ray exec`` twice separately.
```

File: doc/source/autoscaling.rst [hunk 10]
```diff
@@ -240,6 +242,8 @@ The default idle timeout is 5 minutes. This is to prevent excessive node churn w
 Monitoring cluster status
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
+The autoscaler also comes with an online dashboard. To access the dashboard, which is located on the ray head node on port ``8265``. To access it, forward the port to your local machine. You can also call ``ray dashboard`` to do this.
```
Comment (edoakes) [inline doc/source/autoscaling.rst]: The wording of this is sort of weird - ray comes with a dashboard (not the autoscaler) and the dashboard isn't online.

Comment (edoakes) [inline doc/source/autoscaling.rst]:
```suggestion
The autoscaler also comes with an online dashboard. To access the dashboard, which is located on the ray head node on port ``8265``. To access it locally, you'll need to forward the port to your local machine. You can also use the built-in ``ray dashboard`` to do this automatically.
```

File: python/ray/autoscaler/updater.py [hunk 11]
```diff
@@ -44,61 +44,55 @@ def __init__(self, log_prefix, namespace, node_id, auth_config,
...
+            cmd=None,
             timeout=120,
             allocate_tty=False,
             exit_on_fail=False,
             port_forward=None):
-
-        logger.info(self.log_prefix + "Running {}...".format(cmd))
+        if cmd and port_forward:
+            raise Exception(
+                "exec with Kubernetes only can't forward ports and execute"
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]: typo

File: python/ray/autoscaler/updater.py [hunk 12]
```diff
@@ -172,7 +166,15 @@ def get_default_ssh_options(self, connect_timeout):
             ("ControlMaster", "auto"),
             ("ControlPath", "{}/%C".format(self.ssh_control_path)),
             ("ControlPersist", "10s"),
+            # Try less extraneous key pairs
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]:
```suggestion
            # Try fewer extraneous key pairs.
```

File: python/ray/autoscaler/updater.py [hunk 13]
```diff
@@ -172,7 +166,15 @@ def get_default_ssh_options(self, connect_timeout):
...
             ("IdentitiesOnly", "yes"),
+            # Abort if port forwarding fails (instead of just printing to
+            # stderr)
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]:
```suggestion
            # stderr).
```

File: python/ray/autoscaler/updater.py [hunk 14]
```diff
@@ -172,7 +166,15 @@ def get_default_ssh_options(self, connect_timeout):
...
+            ("ExitOnForwardFailure", "yes"),
+            # Quickly kill the connection if network connection breaks (as
+            # opposed to hanging/blocking)
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]:
```suggestion
            # opposed to hanging/blocking).
```

File: python/ray/autoscaler/updater.py [hunk 15]
```diff
@@ -243,20 +245,25 @@ def run(self,
...
 
         final_cmd = ssh + self.get_default_ssh_options(timeout) + [
             "{}@{}".format(self.ssh_user, self.ssh_ip)
-        ] + with_interactive(cmd)
+        ]
+        if cmd:
+            final_cmd += with_interactive(cmd)
+        else:
+            # We do this because `-o ControlMaster` causes the `-N` flag to
+            # still create an interactive shell in some ssh versions
```
Comment (edoakes) [inline python/ray/autoscaler/updater.py]:
```suggestion
            # still create an interactive shell in some ssh versions.
```

File: doc/source/autoscaling.rst [hunk 16]
```diff
@@ -240,6 +242,8 @@ The default idle timeout is 5 minutes. This is to prevent excessive node churn w
 Monitoring cluster status
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
+The ray also comes with an online dashboard. The dashboard is accessible via HTTP on the head node (by default it listens on ``localhost:8265``). To access it locally, you'll need to forward the port to your local machine. You can also use the built-in ``ray dashboard`` to do this automatically.
```
Comment (edoakes) [inline doc/source/autoscaling.rst]:
```suggestion
The ray also comes with an online dashboard. The dashboard is accessible via HTTP on the head node (by default it listens on ``localhost:8265``). To access it locally, you'll need to forward the port to your local machine. You can also use the built-in ``ray dashboard`` command to do this automatically.
```

forgot word

**Review by edoakes — APPROVED** [review]
LGTM!

**Pull request merged**